### ISSUE TITLE
A safe alternative for capturing child process output

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#7171]
There is the top-level method `` `()`` which captures the output of a string command executed in a subshell. The string literal fully supports interpolation, thus it could easily lead to command injection when interpolating user input.

I've just written a docs chapter about this (crystal-lang/crystal-book#304).

In order to avoid command injection, arguments need to be properly escaped. `Process.new` and `Process.run` allow this, but require a few lines of code in order to implement a save equivalent to a simple backticks command.

There is also the top-level method `system` which offers safe arguments, but doesn't capture the output. Instead, it is forwarded to `STDOUT` directly.

There should be method to provide a safe & easy to use alternative for capturing subshell output with the same arguments as `system()` (plus maybe an additional a `shell` argument) but the behaviour and return type of `` `()``.

This *could* be implemented by adding an optional `args` argument to `` `()``, but I don't think that's a good idea because it makes it hard to distinguish the safe operation of this method.

**Comment by ysbaddaden** [comment]
Maybe `output = Process.capture(cmd, args, ...)`?

With all the other arguments except `output` that is hardcoded as `Process::Redirect::Pipe`. Then implementation is a mere:

```crystal
def self.capture(cmd, args, ...)
  Process.run(cmd, args, output: Process::Redirect::Pipe, ..., &.output.gets_to_end)
end
```

**Comment by straight-shoota** [comment]
👍 to `Process.capture`

But I'm wondering about the return type.
Returning a string containing only the standard output might actually be fine for many use cases.
It does not directly expose the error output and exit status of the process though.
Both are available, via the `error` argument and `$?`. So maybe it's fine.

I'm wondering if we could expose the error stream more conveniently.
When running processes in a terminal, you usually get error and output streams merged. But that makes it hard to parse the output. So probably not a good idea.
``::`()`` and `::system()` _inherit_ `STDERR`. That seems very weird.

One option could be to return a tuple containing status, output and error. Or maybe a dedicated type instead of tuple.
But that's all more complexity then, and maybe not necessary?

**Comment by HertzDevil** [comment]
Returning a tuple might make sense if we decide to have a similar macro method superseding the macro backtick (#11456), since `IO` isn't a thing there and `$?` isn't implemented in the macro language yet. The two methods would then be analogous to each other.

**Comment by straight-shoota** [comment]
Another useful feature could be integrated error handling. Maybe that's a step more advanced than a simple capture method. But it seems like a logical evolution and very convenient. We should take it into consideration.
If you have a method that only returns the standard output, it's easy to miss that the command might have failed and you should've handled that.

A very simple implementation of exit-status based error handling:
```cr
output, error = Process.run(cmd, output: :pipe, error: :pipe) do |process|
  {process.output.gets_to_end, process.error.gets_to_end}
end
status = $?
unless status.success?
  raise "Process failed: #{cmd} with status #{status}: #{error}"
end
```
Something similar can be found in [`Compiler#expand_lib_flags`](https://github.com/crystal-lang/crystal/blob/8afd5be1f4c517343aff182e065fd63007fb549e/src/compiler/crystal/compiler.cr#L548-L556), for example.

There are a couple of further examples where a tool like this is re-implemented. `shards` alone has several reusable variations, for example [`Shards::GitResolver#run_in_folder`](https://github.com/crystal-lang/shards/blob/a684407a0815840782f5eef49c19597706ec554f/src/resolvers/git.cr#L433-L457).

Of course this could grow ultimately more complex. Including the full error stream may or may not be a good idea, for example. Maybe some exit statuses are okay to ignore or ask for a more specific exception (e.g. [`Compiler#run_linker`](https://github.com/crystal-lang/crystal/blob/8afd5be1f4c517343aff182e065fd63007fb549e/src/compiler/crystal/compiler.cr#L881-L913)).
Somewhere must be a balance that we could put in a reusable method.

**Comment by ysbaddaden** [comment]
I meant for `capture` to only return STDOUT on purpose — and forgot to mention that it should raise on non success exit code. That's the most usual use-case: you expect the command to be successful and want to grab STDOUT, and to raise an exception otherwise.

If you must handle the exit code without exceptions and/or need to stream/merge both STDERR and STDOUT then you're better off with the regular `.run` and `.new` that give you complete control.

**Comment by straight-shoota** [comment]
It would be really nice if we could provide an alternative that returns captured output and captured error, plus exit status.
Maybe `Process.capture?` indicating that it's a non-raising alternative?

```cr
# Returns captured output.
# Raises if the process exists unsuccessfully.
def Process.capture(command, args, ...) : String

# Returns state, captured output and captured error.
def Process.capture?(command, args, ...) : {Process::Status, String, String}
```

We could integrate the signature of `capture?` into the macro language. Although there the question mark is a bit odd if there's no alternative.

**Comment by straight-shoota** [comment]
An alternative would be `capture!` for the raising variant and `capture` for the one that returns a tuple.
This might not be ideal because the blank name is actually a more advanced method and requires more work than the other one which despite the exclamation mark is arguably the safer variant.

**Comment by straight-shoota** [comment]
Another use case: When a process is not expected to print any output, but we still want to raise if it's unsuccessful.
There's no need to capture stdout then, so it might be convenient to have a variant of `Process.run` raises if the exit status is not successful.
But maybe it's fine to just use `Process.capture` for that, despite the little unnecessary overhead for capturing?

This would work well as `Process.run!` and `Process.capture!`, though.

**Comment by HertzDevil** [comment]
> there the question mark is a bit odd if there's no alternative

This is fine since the macro methods are less strict about nils (e.g. `ArrayLiteral#[]` and `env` simply return nil without a question mark variant).

**Comment by ysbaddaden** [comment]
After drafting the method in #16630, we might want to add:

- `Process.run!` that raises when `status.success?` is false;
- `Process.capture!` that also raises (instead of `Process.capture`).

**Comment by Fryguy** [comment]
This might be scope-creep, but I like the idea of capture as a replacement/alternative for backticks, but it feels incomplete. It's also not clear on its face if it only captures stdout, or also capture stderr (or both merged)? I do see that discussed above, but I'm trying put myself in a new user's shoes, and it feels unclear to me.

I'm wondering if we should take inspiration from [Ruby's open3 module](https://github.com/ruby/open3?tab=readme-ov-file#usage), and have separate methods like capture/capture2/capture2e (or perhaps params driving things through a single capture method so the names aren't so ugly).

**Comment by ysbaddaden** [comment]
There's already `Process.run(&)` to stream the 3 standard descriptors. So far, there's no use case for capturing both stdout _and_ stderr simply, only stdout is needed. I don't see any incentive to bring something like `#capture2` yet (also: awful name).

**Comment by straight-shoota** [comment]
We could consider capturing stderr to put it into the exception as mentioned in https://github.com/crystal-lang/crystal/issues/7171#issuecomment-3806976341
I figure this might be quite useful.

Probably not as part of the error message, though. But it could be a property on a dedicated error type (alongside `status`), which would easily make relevant information about the failed process available in an error handler.

**Comment by ysbaddaden** [comment]
Let's note that capturing both stdout and stderr requires to spawn at least one fiber to read async from one of the stream, because only reading one stream then the other stream can lead the sub-process to block trying to write on the other stream (oops).

```crystal
def capture(...)
  output = error = nil
  wg = WaitGroup.new

  process = Process.new(..., output: Redirect::Pipe, error: Redirect::Pipe)

  wg.spawn do
    error = process.error.gets_to_end
  rescue
    # silence exceptions
  end

  output = process.output.gets_to_end

  process.wait
  wg.wait

  raise Error.new(error: error) unless process.status.success?

  output
end
```

Takeaway:

1. Given the complexity and trickiness, having a capture method that captures both streams would be valuable.

2. Raising means that stderr is necessarily an error stream, but it may be something entirely different; I think returning a `{status, output, error}` triplet would be much more adequate and flexible.

3. Given that the added WaitGroup, Fiber and String::Builder (error stream) are expected to be thrown away in far too many cases, I think we still want a basic alternative that focuses on the happy path: capture/return stdout and a mere `raise "failed"`.

So, maybe:

- `Process.capture(...) : {Process::Status, String, String}`
- `Process.capture!(...) : String`

**Comment by straight-shoota** [comment]
I think the most basic variant should indeed capture stderr, raise on unsuccessful status and make stderr available in the exception.

That might not be the most efficient solution, but it's the most convenient and useful. That's what we should offer as the standard behaviour.

We can offer a more efficient alternative that does not capture stderr. Though I'm wondering if the difference is even that relevant in contrast to the effort of spawning a process and assuming the error stream doesn't see much activity.

It might be relatively easy to offer this alternative mode by exposing `error` as a parameter with `:pipe` as default value. Passing any other value (such as `error: :close`) avoids capturing stderr.

FWIW, the equivalent in Go, [`Cmd.Output()`](https://pkg.go.dev/os/exec#Cmd.Output) always captures stderr and exposes it in the error value. The captured amount is limited however, preserving only the first 32k and last 32k bytes (see [`os.exec.ExitError`](https://pkg.go.dev/os/exec#ExitError)).

**Comment by straight-shoota** [comment]
We should consider a couple different modes of behaviour. That doesn't mean all of them should be available as methods. Certainly not right away. Their practical value varies. I marked the most relevant ones in bold text.

But it makes sense to look at the full picture and take extensibility into account.

### 1. Status-only processes

These are commands used as predicates or side effects.

| variant | successful | exit failure | start failure | example |
|-|-|-|-|-|
| 1a | `true` / `nil` | raises (command, status, stderr) | raises (command) |
| **1b** | `true` | `false` | raises (command) | `Process.run(...).success?` |
| **1c** | `status` | `status` | raises (command) | `Process.run(...)` |
| **1d** | `status` | `status`| `nil` | `Process.run?(...)` (#9896) |

* 1b is very similar to the existing `::system`. `::system` currently only exposes parameters `command` and `args` and implicitly inherits both stderr and stdout. It could be called `Process.run?` but that would collide with #9896.
* 1c already exists as `Process.run`

### 2. Only stdout matters

We're only interested in stdout and whether the command was successful.

| variant | successful | exit failure | start failure | example |
|-|-|-|-|-|
| **2a** | `stdout` | raises (command, status, stdout, stderr) | raises (command) | `Process.capture(...)` |
| 2b | `stdout` | `nil` | raises (command) / `nil` (?)  | `Process.capture?(...)` |
| 2c | `{status, stdout}` | raises (command) | `{status, stdout}` |

* 2c is easy to replace by 3c

### 3. Stdout and stderr matter

Returns the full captured information.

| variant | successful | exit failure | start failure | example |
|-|-|-|-|-|
| 3a | `{stdout, stderr}` | raises (command, status, stdout, stderr) | raises (command) |
| 3b | `{stdout, stderr}` | `nil` | raises (command) |
| **3c** | `{status, stdout, stderr}` | `{status, stdout, stderr}` | raises (command) | `Process.capture_result` (?) |
| 3d | `{status, stdout, stderr}` | `{status, stdout, stderr}` | nil | `Process.capture_result?` (?) |

* Instead of an actual tuple, a dedicated result type (`Process::Result`?) might be more appropriate for this.
